### PR TITLE
Add access group to key storage

### PR DESCRIPTION
**Pull request by symorton**
**Problem:**
There was a very very rare error where the keys in default access group were getting deleted by app. 

**Solution:**
Our solution is to save all keys related to VC under a configurable access group.
**We also changed the SDK initialization external api to make it more clear.** 

**Validation:**
All tests pass, and works end to end.

**Type of change:**
- [X] Feature work
- [X] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

**Risk**:
- [X] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [ ] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

**Work Item links**:
Please include here links for this work item, or deferred work, or related work. E.g. if the refactoring is too big to fit in this PR, or the localized strings need to be updated later, please link the TODO work items here.

**Documentation Links**:
Please include here links to any related background documentation for this PR.

### CODE REVIEW

File: VCCrypto/VCCrypto/Keys/Random32BytesSecret.swift [hunk 1]
```diff
@@ -58,4 +61,47 @@ final class Random32BytesSecret: Secret {
...
+        } catch {
+            return false
+        }
+    }
+    
+    func migrateKey(fromAccessGroup oldAccessGroup: String?) throws {
+        
+        /// If old access group is equal to new access group, no action required.
+        if oldAccessGroup == accessGroup
+        {
```
Comment (dangodb) [inline VCCrypto/VCCrypto/Keys/Random32BytesSecret.swift:78]: nit: style

File: VCCrypto/VCCrypto/Keys/Random32BytesSecret.swift [hunk 2]
```diff
@@ -58,4 +61,47 @@ final class Random32BytesSecret: Secret {
...
+    func isValidKey() -> Bool {
+        do {
+            _ = try self.store.getSecret(id: id, itemTypeCode: Random32BytesSecret.itemTypeCode, accessGroup: accessGroup)
+            return true
+        } catch {
+            return false
+        }
+    }
+    
+    func migrateKey(fromAccessGroup oldAccessGroup: String?) throws {
```
Comment (dangodb) [inline VCCrypto/VCCrypto/Keys/Random32BytesSecret.swift:74]: nit: currentAccessGroup

File: VCEntities/VCEntities/identifier/KeyContainer.swift [hunk 3]
```diff
@@ -25,4 +25,13 @@ public struct KeyContainer {
     public func getId() -> UUID {
         return keyReference.id
     }
+    
+    public func isValidKey() -> Bool {
+        return keyReference.isValidKey()
+    }
+    
+    /// Migrate key from old access group to new one set in sdk config.
+    public func migrateKey(fromAccessGroup oldAccessGroup: String?) throws {
```
Comment (dangodb) [inline VCEntities/VCEntities/identifier/KeyContainer.swift:34]: same

File: VCEntities/VCEntities/logging/VCSDKLog.swift [hunk 4]
```diff
@@ -93,7 +93,7 @@ public struct VCSDKLog {
         }
     }
     
-    func event(name: String, properties: [String: String]? = nil, measurements: [String: NSNumber]? = nil) {
+    public func event(name: String, properties: [String: String]? = nil, measurements: [String: NSNumber]? = nil) {
```
Comment (dangodb) [inline VCEntities/VCEntities/logging/VCSDKLog.swift:96]: Why did we have to change it to public?

File: VCEntities/VCEntities/util/VCSDKConfiguration.swift [hunk 5]
```diff
@@ -7,11 +7,11 @@ public struct VCSDKConfiguration {
     
     public static var sharedInstance = VCSDKConfiguration()
     
-    public var userAgentInfo: String = ""
+    public private(set) var accessGroupIdentifier: String? = nil
     
     private init() {}
     
-    public mutating func setUserAgentInfo(with info: String) {
```
Comment (dangodb) [inline VCEntities/VCEntities/util/VCSDKConfiguration.swift:14]: Not using it?

File: VCServices/VCServices/IdentifierService.swift [hunk 6]
```diff
@@ -51,4 +51,21 @@ public class IdentifierService {
         sdkLog.logVerbose(message: "Created Identifier with alias:\(identifier.alias)")
         return identifier
     }
+    
+    /// updates access group for keys if it needs to be updated.
+    public func migrateKeys(fromAccessGroup oldAccessGroup: String?) throws {
```
Comment (dangodb) [inline VCServices/VCServices/IdentifierService.swift:56]: same

**Review by dangodb — CHANGES_REQUESTED** [review]

File: VCEntities/VCEntities/util/VCSDKConfiguration.swift (hunk 5 above)
Comment (symorton) [reply VCEntities/VCEntities/util/VCSDKConfiguration.swift:14]: nope

File: VCCrypto/VCCrypto/KeychainSecretStore.swift [hunk 7]
```diff
@@ -66,15 +75,58 @@ struct KeychainSecretStore : SecretStoring {
...
     }
+    
+    /// Delete the secret from keychain
+    /// Note: the value passed in will be zeroed out after the secret is deleted
+    /// - Parameters:
+    ///   - id: The Id of the secret
+    ///   - itemTypeCode: The secret type code (4 chars)
+    ///   - accessGroup: The access group of the secret.
+    ///   - value: The value of the secret
+    func deleteSecret(id: UUID, itemTypeCode: String, accessGroup: String? = nil, value: inout Data) throws {
```
[the remark this answers is not in the document]
Comment (symorton) [reply VCCrypto/VCCrypto/KeychainSecretStore.swift:102]: sounds good.

File: VCEntities/VCEntities/logging/VCSDKLog.swift (hunk 4 above)
Comment (symorton) [reply VCEntities/VCEntities/logging/VCSDKLog.swift:96]: reverting

File: VCServices/VCServices/VerifiableCredentialSDK.swift [hunk 8]
```diff
@@ -5,26 +5,53 @@
...
         } catch {
+            
+            /// If unable to fetch master identifier, create a new one.
             VCSDKLog.sharedInstance.logWarning(message: "Failed to fetch master identifier with: \(String(describing: error))")
-            _ = try identifierService.createAndSaveIdentifier(forId: VCEntitiesConstants.MASTER_ID, andRelyingParty: VCEntitiesConstants.MASTER_ID)
-            return true
+            return createNewIdentifier()
+            
+        }
+
```
Comment (dangodb) [inline VCServices/VCServices/VerifiableCredentialSDK.swift:43]: Shouldn't we check here to see if the key can be read and fail if it's not? Then for the caller it's cleaner if the init succeed you know you're in a good state and if not then you can handle it differently based on the error returned

**Review by dangodb — CHANGES_REQUESTED** [review]

File: VCServices/VCServices/VerifiableCredentialSDK.swift (hunk 8 above)
Comment (symorton) [reply VCServices/VCServices/VerifiableCredentialSDK.swift:43]: oh yes thats what we wanted, good catch

**Review by dangodb — APPROVED** [review]

**Pull request merged**